### PR TITLE
Completing google authentication

### PR DESCRIPTION
**Pull request by yatharth77**
Screenshot 2021-02-02 at 1 21 16 AM

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -15,6 +15,19 @@
...
+
+    <!-- Latest compiled and minified CSS -->
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
+
+    <!-- jQuery library -->
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+
+    <!-- Latest compiled JavaScript -->
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
+    <script type="text/javascript" src="http://ajax.microsoft.com/ajax/jquery/jquery-1.6.2.min.js"></script>
```
Comment (tushargupta191) [inline public/index.html:30]: Lets discuss what all scripts are required here

File: src/App.css [hunk 2]
```diff
@@ -1,38 +1,102 @@
...
+  display: flex;
+  width: 100%;
+  align-items: stretch;
 }
 
-@media (prefers-reduced-motion: no-preference) {
-  .App-logo {
-    animation: App-logo-spin infinite 20s linear;
-  }
+#sidebar {
```
Comment (tushargupta191) [inline src/App.css:14]: We can probably stick to just using classes

File: src/App.css [hunk 3]
```diff
@@ -1,38 +1,102 @@
...
+
+.dropdown-toggle::after {
+  display: block;
+  position: absolute;
+  top: 50%;
+  right: 20px;
+  transform: translateY(-50%);
+}
+
+@import "https://fonts.googleapis.com/css?family=Poppins:300,400,500,600,700";
```
Comment (tushargupta191) [inline src/App.css:36]: Import can be moved to the top

File: src/utils/apiManager.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+export class ApiManager {
+    static BASE_URL = 'https://gmail.googleapis.com/gmail/v1/users/';
+    constructor(user, accessToken){
+        this.user = user;
+        this.accessToken = accessToken;
+        this.objectName = "";
```
Comment (tushargupta191) [inline src/utils/apiManager.js:6]: We can remove objectName and objectId from here?
Also, here user is just username right? If yes, we can probably rename it

File: src/utils/apiManager.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        this.objectId = "";
+    }
+
+    fetchAPI(objectName, objectId) {
+        return new Promise((resolve, reject) => {
+            var xhr = new XMLHttpRequest()
```
Comment (tushargupta191) [inline src/utils/apiManager.js:12]: Lets prefer using let/const.

Also, not for now, but we should try implementing the api-manager using axios, since that is what we are currently using. It will help you get a rough idea on how that looks like.

Not urgent

File: src/utils/refreshToken.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export const refreshTokenSetup = (res) => {
+    let refreshTiming = (res.tokenObj.expires_in || 3600 - 5 * 60 ) * 1000;
```
Comment (tushargupta191) [inline src/utils/refreshToken.js:2]: Lets move the 3600-5*60 to a constant as done with api-manager.

File: src/components/Sidebar.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import React, { Component } from 'react'
+import Dexie from 'dexie'
+
+class Sidebar extends Component {
+    constructor(props){
```
Comment (tushargupta191) [inline src/components/Sidebar.js]: This constructor should not be required

File: src/components/Sidebar.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        super(props);
+    }
+
+    handleLabel = async (labelName, setThreadDetails) => {
```
Comment (tushargupta191) [inline src/components/Sidebar.js]: Lets rename this function so that it is more self-explanatory

File: src/components/Sidebar.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        labelName = labelName.toUpperCase();
+        const db = await new Dexie("flockdev07@gmail").open()
```
Comment (tushargupta191) [inline src/components/Sidebar.js]: Not for now,
but we should get this value 'flockdev@gmail.com' from the user's object and not hard-code it.

File: src/components/Sidebar.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        const dataTable = db.table("threads");
+        dataTable.filter(function(thread){
+            return (thread.labels.includes(labelName));
+        }).toArray(data => {
```
Comment (tushargupta191) [inline src/components/Sidebar.js]: Can simplify the inner function to,

dataTable.filter(thread => thread.lables.includes(labelName)).toArray....

File: src/components/Sidebar.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                setThreadDetails(data);
+        });
+    }
+    
+    processLabels = () => {
```
Comment (tushargupta191) [inline src/components/Sidebar.js]: Can probably rename this to renderlabels

File: src/App.js [hunk 12]
```diff
@@ -1,25 +1,78 @@
...
+// import Logout from './components/Logout'
+import './App.css'
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+        signedIn: false,
+        accessToken: "",
+        labels: [],
```
Comment (tushargupta191) [inline src/App.js:15]: We should move labels, accessToken and threadDetails out for App's state. 
A part of this will be achieved with liveQuery.

File: src/App.js [hunk 13]
```diff
@@ -1,25 +1,78 @@
...
+
+  signOut = () => {
+    this.setState({ signedIn: false});
+  }
+
+  signIn = () => {
+    this.setState({ signedIn: true});
+  }
+
+  setAccessToken = (access_token) => {
```
Comment (tushargupta191) [inline src/App.js:29]: Instead of having a signedIn state, can we determine user's signed in status by checking of access_token exists or not?

File: src/App.js [hunk 14]
```diff
@@ -1,25 +1,78 @@
...
+                    accessToken={this.state.accessToken}
+                    labels={this.state.labels}
+                    setThreadDetails={this.setThreadDetails}
+                  />
+                </div>
+              </nav>
+
+            <div className="login-div">
+              { this.state.signedIn 
+                ? <h1>Welcome, you are signed <hr/> <LoadMessages threadDetails={this.state.threadDetails}/></h1>
```
Comment (tushargupta191) [inline src/App.js]: Lets rename LoadMessages to probably loadThreads.
Also, let LoadThreads be responsible for fetching threads from db rather than App.js doing it

File: src/components/LoadMessages.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import React, { Component } from 'react'
+import parse from 'html-react-parser';
+import '../style/style.css'
+
+class LoadMessages extends Component {
+    constructor(props) {
```
Comment (tushargupta191) [inline src/components/LoadMessages.js]: constructor should not be required.
Also, lets rename this file

File: src/components/LoadMessages.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    processThreads = () => {
+        const threadDetailsArray = this.props.threadDetails;
+        return (
+            <div className="list-group">
+                {
+                    threadDetailsArray.map((thread, threadIndex) => {
+                        const messages = thread.messages;
+                        return (
+                            <div key={threadIndex}>
+                                <a href="#" className="list-group-header">Message Thread {threadIndex+1}</a>
```
Comment (tushargupta191) [inline src/components/LoadMessages.js:15]: Why are we using 'a' tag here?

File: src/components/Login.js [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    
+    constructor(props){
+        super(props);
+        this.state = {
+            dbInstance: null,
+        }
+    }
+
+    connectIndexedDB = () => {
+        var db = new Dexie("flockdev07@gmail");
```
Comment (tushargupta191) [inline src/components/Login.js:12]: Lets use let/const
Similar to api-manager, lets make a db-manager, which is responsible for creating a dexie instance and takes care of versioning
Lets try using flockdev07@gmail.com from user's details rather than hardcoding

File: src/components/Login.js [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        db.version(1).stores({
+            labels: 'id',
+            threads: 'id, labels',
+            messages: 'id',
+        })
+        this.setState({ dbInstance: db });
+    }
+
+    serachDB = async (dbName, tableName) => {
+        const exist = await Dexie.exists(dbName);
```
Comment (tushargupta191) [inline src/components/Login.js:28]: This function can also be moved to the db-manager

File: src/components/Login.js [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    //     const db  = this.state.dbInstance;
+    //     console.log(db, db.threads);
+    //     db.threads.filter(function(thread){
+    //         return (thread.labels.includes(category));
+    //     }).each(function (thread){
+    //         console.log(thread);
+    //     })
+    // }
+
+    handleResponse = async (response) => {
```
Comment (tushargupta191) [inline src/components/Login.js:44]: can rename this more aptly

File: src/components/Login.js [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+            apiManager.fetchAPI("labels", "", this.props.accessToken).then((labelJson) => {
+                this.props.setLabels(labelJson.labels);
+                labelJson.labels.map((value) => {
+                    this.state.dbInstance.labels.put({ ...value })
+                })
+            })
+        }
+
+        const threadData = await this.serachDB("flockdev07@gmail", "threads");
+        if(threadData){
```
Comment (tushargupta191) [inline src/components/Login.js:69]: Lets move this to load-thread file.

File: src/components/Logout.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    handleResponse = () => {
+        this.props.setSignedInState({ signedIn: false });
+    }
+    render(){
+        return(
+        <div>
+            <GoogleLogout 
+            clientId={'1025950630557-ir2p4fp2tnkl0co0il6jgee1ikhdd054.apps.googleusercontent.com'}
```
Comment (tushargupta191) [inline src/components/Logout.js:16]: This can be served from the constant file
